File: MatlabOctaveCodeSnippets/snip_systems_overlapAdd.m
x=1:1000; %infinite duration (or "long") input signal
h=ones(1,3); %non-zero samples of finite-length impulse response
Nh=length(h); %number of impulse response non-zero samples
Nb=5; %block (segment) length
Nfft=2^nextpow2(Nh+Nb-1); %choose a power of 2 FFT size
Nx = length(x); %number of input samples
H = fft(h,Nfft); %pre-compute impulse response DFT, with zero-padding
beginIndex = 1; %initialize index for first sample of current block
y = zeros(1,Nh+Nx-1); %pre-allocate space for convolution output
while beginIndex <= Nx %loop over all blocks
    endSample = min(beginIndex+Nb-1,Nx);%last sample of current block
    Xblock = fft(x(beginIndex:endSample),Nfft); %DFT of block
    yblock = ifft(Xblock.*H,Nfft); %get circular convolution result
    outputIndex  = min(beginIndex+Nfft-1,Nh+Nx-1); %auxiliary variab.
    y(beginIndex:outputIndex) = y(beginIndex:outputIndex) + ...
        yblock(1:outputIndex-beginIndex+1); %add parcial result
    beginIndex = beginIndex+Nb; %shift begin of block
end
stem(y-conv(x,h)) %compare the error with result from conv

File: PythonCodeSnippets/snip_systems_overlapAdd.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.fft import fft, ifft

# Infinite duration (or "long") input signal
x = np.arange(1, 1001)

# Non-zero samples of finite-length impulse response
h = np.ones(3)

# Number of impulse response non-zero samples
Nh = len(h)

# Block (segment) length
Nb = 5

# Choose a power of 2 FFT size
Nfft = 2 ** np.ceil(np.log2(Nh + Nb - 1)).astype(int)

# Number of input samples
Nx = len(x)

# Pre-compute impulse response DFT, with zero-padding
H = fft(h, Nfft)

# Initialize index for first sample of current block
begin_index = 0

# Pre-allocate space for convolution output
y = np.zeros(Nh + Nx - 1)

# Loop over all blocks
while begin_index < Nx:
    end_sample = min(begin_index + Nb, Nx)  # Last sample of current block
    Xblock = fft(x[begin_index:end_sample], Nfft)  # DFT of block
    yblock = ifft(Xblock * H, Nfft)  # Get circular convolution result
    output_index = min(begin_index + Nfft, Nh + Nx - 1)  # Auxiliary variable
    # Add partial result, ensuring to only use the real part of the complex numbers
    y[begin_index:output_index] += yblock[:output_index - begin_index].real
    begin_index += Nb  # Shift beginning of block


# Compare the error with result from conv
error = y - np.convolve(x, h, mode='full')
plt.stem(error)
plt.show()